contracts/src/mixins/ERC4337Mixin.sol

```diff
@@ -0,0 +1,244 @@
+    // ------------------------------------------------------------------
+    // 4337 functions
+    // ------------------------------------------------------------------
+
+    /**
+     * @notice The address of the EntryPoint contract supported by this module.
+     */
+    address public immutable SUPPORTED_ENTRYPOINT;
+
+    constructor(address _entryPoint) {
+        SUPPORTED_ENTRYPOINT = _entryPoint;
+    }
+
+    /**
+     * Return the account nonce.
+     * This method returns the next sequential nonce.
+     * For a nonce of a specific key, use `entrypoint.getNonce(account, key)`
+     */
+    function getNonce(address signer) public view virtual returns (uint256) {
+        return
+            IEntryPoint(SUPPORTED_ENTRYPOINT).getNonce(
+                address(this),
+                uint192(uint160(signer))
+            );
+    }
+
+    function validateUserOp(
+        PackedUserOperation calldata userOp,
+        bytes32,
+        uint256 missingAccountFunds
+    ) external override returns (uint256 validationData) {
+        // Assumption:
+        //   - UserOp signature is SafeTx signature
+        // Requirements:
+        //   - [x] Check entrypoint
+        //   - [x] No paymaster support
+        // Steps:
+        //   - [x] decode callData
+        //   - [x] recover signer from SafeTx signature
+        //   - [x] check that signer has not submitted signature
+        //   - [x] check that nonce key is signer address
+        //   - [x] pay missing funds
+        //   - [ ] usage checks -> set validation data
+
+        require(
+            msg.sender == SUPPORTED_ENTRYPOINT,
+            InvalidEntryPoint(msg.sender)
+        );
+        require(
+            userOp.paymasterAndData.length == 0,
+            InvalidUserOpPaymasterAndData()
+        );
+        require(userOp.signature.length == 65, InvalidECDSASignatureLength());
+
+        require(
+            bytes4(userOp.callData) == this.storeTransaction.selector,
+            InvalidTarget(bytes4(userOp.callData))
+        );
+        (
+            bytes32 safeTxHash,
+            address signer,
+            bytes32 r,
+            bytes32 vs
+        ) = _verifySafeTxHash(userOp.callData[4:]);
+
+        // --- DUPLICATE TRANSACTION SIGNATURE CHECK ---
+        // Revert if this signer has already submitted *any* signature for this *exact* safeTxHash
+        require(
+            !_signerSignedTx(safeTxHash, signer),
+            SignerAlreadySignedTransaction(signer, safeTxHash)
+        );
+
+        _verifySignature(safeTxHash, userOp.signature, signer, r, vs);
+
+        uint256 nonce = getNonce(signer);
+        require(userOp.nonce == nonce, UnexpectedNonce(nonce));
+
+        // We trust the entry point to set the correct prefund value, based on the operation params
+        // We need to perform this even if the signature is not valid, else the simulation function of the entry point will not work.
+        if (missingAccountFunds != 0) {
```

Comment:
Currently the harbour contract doesn't have a payable method, so we could argue to remove this branch and require that `missingAccountFunds` should always be 0

Reply:
One thing that we can consider adding would be a way to deposit per signer. Right now, there is only the global "balance" and anyone can drain it with valid Safe transaction signatures.

Reply:
I'm indifferent on whether or not the code is left in for now, so long as we have a solution to "how to fund the harbour, without needing to setup a personal paymaster for it".